pkg/resources/grant_application_role.go

```diff
@@ -79,54 +86,91 @@ func GrantApplicationRole() *schema.Resource {
 		snowflakeResourceID = helpers.EncodeSnowflakeID(applicationRoleIdentifier.FullyQualifiedName(), sdk.ObjectTypeApplication.String(), applicationIdentifier.FullyQualifiedName())
 		req := sdk.NewGrantApplicationRoleRequest(applicationRoleIdentifier).WithTo(*sdk.NewKindOfRoleRequest().WithApplicationName(&applicationIdentifier))
 		if err := client.ApplicationRoles.Grant(ctx, req); err != nil {
-			return err
+			return diag.FromErr(err)
 		}
 	}
 	d.SetId(snowflakeResourceID)
-	return ReadGrantApplicationRole(d, meta)
+	return ReadContextGrantApplicationRole(ctx, d, meta)
 }
 
-func ReadGrantApplicationRole(d *schema.ResourceData, meta interface{}) error {
+func ReadContextGrantApplicationRole(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(*provider.Context).Client
 	parts := strings.Split(d.Id(), helpers.IDDelimiter)
 	applicationRoleName := parts[0]
 	applicationRoleIdentifier := sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(applicationRoleName)
-	objectType := parts[1]
+	objectTypeString := parts[1]
+	if objectTypeString == "ACCOUNT_ROLE" {
+		objectTypeString = "ROLE"
+	}
+
+	// first check if either the target account role or application exists
 	targetIdentifier := parts[2]
-	ctx := context.Background()
+	objectType := sdk.ObjectType(objectTypeString)
+	switch objectType {
+	case sdk.ObjectTypeRole:
+		{
+			if _, err := client.Roles.ShowByID(ctx, sdk.NewShowByIdRoleRequest(sdk.NewAccountObjectIdentifierFromFullyQualifiedName(targetIdentifier))); err != nil && errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
+				d.SetId("")
+				return diag.Diagnostics{
+					diag.Diagnostic{
+						Severity: diag.Warning,
+						Summary:  "Failed to retrieve account role. Marking the resource as removed.",
+						Detail:   fmt.Sprintf("Id: %s", d.Id()),
+					},
+				}
+			}
+		}
+	case sdk.ObjectTypeApplication:
+		if _, err := client.Applications.ShowByID(ctx, sdk.NewAccountObjectIdentifierFromFullyQualifiedName(targetIdentifier)); err != nil && errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
+			d.SetId("")
+			return diag.Diagnostics{
+				diag.Diagnostic{
+					Severity: diag.Warning,
+					Summary:  "Failed to retrieve application. Marking the resource as removed.",
+					Detail:   fmt.Sprintf("Id: %s", d.Id()),
+				},
+			}
+		}
+	}
+	// then check if application role exists
 	grants, err := client.Grants.Show(ctx, &sdk.ShowGrantOptions{
 		Of: &sdk.ShowGrantsOf{
 			ApplicationRole: applicationRoleIdentifier,
 		},
 	})
-	if err != nil {
-		log.Printf("[DEBUG] application role (%s) not found", applicationRoleIdentifier.FullyQualifiedName())
+	if err != nil && errors.Is(err, sdk.ErrObjectNotExistOrAuthorized) {
 		d.SetId("")
-		return nil
+		return diag.Diagnostics{
+			diag.Diagnostic{
+				Severity: diag.Warning,
+				Summary:  "Failed to retrieve application role. Marking the resource as removed.",
+				Detail:   fmt.Sprintf("Id: %s", d.Id()),
+			},
+		}
 	}
```

Comment:
shouldn't we also check the err even if it's not the `ErrObjectNotExistOrAuthorized`? Right now, we discard the error, but maybe in that case we should return an error with `err` as additional context.

Reply:
Fixed.